lib/liquid/block_body.rb

```diff
@@ -70,49 +70,57 @@ def render(context)
       output = []
       context.resource_limits.render_score += @nodelist.length
 
-      @nodelist.each do |token|
-        # Break out if we have any unhanded interrupts.
+      idx = 0
+      while node = @nodelist[idx]
```

Comment:
> A while loop is faster than iterating with #each.

This statement surprised me, so I created this microbenchmark

```ruby
require 'benchmark'

N = 100000
LIST = (0..1000).to_a

Benchmark.bmbm do |x|
  x.report("while") do
    N.times do
      LIST.each do |item|
      end
    end
  end
  x.report("each") do
    N.times do
      idx = 0
      while item = LIST[idx]
        idx += 1
      end
    end
  end
end
```

the results were

```
            user     system      total        real
while   3.470000   0.010000   3.480000 (  3.480269)
each    2.230000   0.000000   2.230000 (  2.232083)
```

so I think this change might have made things worse, even if the overall benchmark improved from the other changes in this PR.

Reply:
When running your benchmark I get the same result.

But when running the liquid benchmark `while` appears to be faster than `each`. Here are three runs using `each`:

```
Run 1)
              parse:     41.466  (± 2.4%) i/s -    416.000  in  10.039060s
             render:     77.517  (± 3.9%) i/s -    777.000  in  10.036251s
     parse & render:     25.457  (± 3.9%) i/s -    256.000  in  10.060507s

Run 2)
              parse:     41.188  (± 0.0%) i/s -    412.000  in  10.004006s
             render:     77.819  (± 2.6%) i/s -    784.000  in  10.081885s
     parse & render:     25.772  (± 3.9%) i/s -    258.000  in  10.015315s

Run 3)
              parse:     41.479  (± 0.0%) i/s -    416.000  in  10.030267s
             render:     76.804  (± 6.5%) i/s -    770.000  in  10.071492s
     parse & render:     24.713  (± 4.0%) i/s -    248.000  in  10.059831s
```

Maybe one of you can check that too.

Reply:
Interesting.  I tried with only the while loop change it did improve performance on the benchmark.